src/wallet/rpcwallet.cpp

```diff
@@ -4370,62 +4370,131 @@ void AddKeypathToMap(const CWallet* pwallet, const CKeyID& keyID, std::map<CPubK
 
-        MutableTransactionSignatureCreator creator(psbtx.tx.get_ptr(), 0, out.nValue.GetAmount(), 1);
-        ProduceSignature(HidingSigningProvider(pwallet, true, !bip32derivs), creator, out.scriptPubKey, sigdata);
-        psbt_out.FromSignatureData(sigdata);
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+        return NullUniValue;
     }
-    return complete;
+
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 4)
+        throw std::runtime_error(
+            RPCHelpMan{"walletsignpsbt",
+                "\nSign all PSBT iputs that we can sign for.\n"
+                + HelpRequiringPassphrase(pwallet) + "\n",
+                {
+                    {"psbt", RPCArg::Type::STR, RPCArg::Optional::NO, "The transaction base64 string"},
+                    {"sighashtype", RPCArg::Type::STR, /* default */ "ALL", "The signature hash type to sign with if not specified by the PSBT. Must be one of\n"
+                        "       \"ALL\"\n"
+                        "       \"NONE\"\n"
+                        "       \"SINGLE\"\n"
+                        "       \"ALL|ANYONECANPAY\"\n"
+                        "       \"NONE|ANYONECANPAY\"\n"
+                        "       \"SINGLE|ANYONECANPAY\""},
+                    {"imbalance_ok", RPCArg::Type::BOOL, /* default */ "false", "Sign even if the transaction amounts do not balance"},
+                },
+                RPCResult{
+                    "{\n"
+                    "  \"psbt\" : \"value\",          (string) The base64-encoded partially signed transaction\n"
+                    "  \"complete\" : true|false,   (boolean) If the transaction has a complete set of signatures\n"
+                    "}\n"
+                },
+                RPCExamples{
+                    HelpExampleCli("walletprocesspsbt", "\"psbt\"")
+                    + HelpExampleRpc("walletprocesspsbt", "\"psbt\"")
+                }
+            }.ToString()
+        );
+
+    RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VSTR, UniValue::VBOOL});
+
+    // Unserialize the transaction
+    PartiallySignedTransaction psbtx;
+    std::string error;
+    if (!DecodeBase64PSBT(psbtx, request.params[0].get_str(), error)) {
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, strprintf("TX decode failed %s", error));
+    }
+
+    // Get the sighash type
+    int nHashType = ParseSighashString(request.params[1]);
+    bool imbalance_ok = request.params[2].isNull() ? false : request.params[2].get_bool();
+
+    bool complete;
+    const TransactionError err = SignPSBT(pwallet, psbtx, complete, nHashType, true, imbalance_ok);
+    if (err != TransactionError::OK) {
+        throw JSONRPCTransactionError(err);
+    }
+
+    UniValue result(UniValue::VOBJ);
+    result.pushKV("psbt", EncodePSBT(psbtx));
+    result.pushKV("complete", complete);
+
+    return result;
 }
 
 UniValue walletprocesspsbt(const JSONRPCRequest& request)
 {
+    if (!g_con_elementsmode)
```

Comment:
In !g_con_elementsmode, a bunch of psbt-related stuff is broken now. I don't know that I really trust any of the code that I fiddled with to operate in non-elements-mode. Certainly none of it has been tested there. So we had discussed just gating everything on elementsmode being on, since that's the only mode that Elements is really meant to be operated in anyway.

This RPC is noted in its help to be deprecated (at least, implicitly via the recommendation to use other RPCs instead.) I could give a stronger deprecation warning, or just remove it entirely. The sequence of walletfillpsbtdata / blindpsbt / walletsignpsbt is a complete replacement for it. It will still work in elementmode, if the transaction to be processed does not have any blinded outputs.